state.go

```diff
@@ -557,6 +561,103 @@ func (s *State) Channel(channelID string) (*Channel, error) {
 }
 
+// ThreadMemberUpdate updates the ThreadMember object for the user, in a specific thread.
+func (s *State) ThreadMemberUpdate(t *ThreadMember) error {
+	if s == nil {
+		return ErrNilState
+	}
+
+	// If the channel exists in the state, update the thread member of that channel
+	if channel, ok := s.channelMap[t.ID]; ok {
+		s.Lock()
+		defer s.Unlock()
+
+		channel.Member = t
+		return nil
+	}
+
+	return ErrStateNotFound
+}
+
+// ThreadMembersUpdate updates the member object of the thread. If the user was removed
+// from the thread, it will set the ThreadMember object in the state to nil.
+func (s *State) ThreadMembersUpdate(t *ThreadMembersUpdate) error {
+	if s == nil {
+		return ErrNilState
+	}
+
+	s.Lock()
+	defer s.Unlock()
+
+	if thread, ok := s.channelMap[t.ID]; ok {
+		for _, memberID := range t.RemovedMemberIDs {
+			if memberID == s.User.ID {
+				// Remove the thread member object from the channel, since we no longer have access to it.
+				thread.Member = nil
+				break
+			}
+		}
+	}
+
+	return ErrStateNotFound
+}
+
+// ThreadListSync syncs the new threads and thread member objects made available
+// after a permission change that allows the client to see more channels.
+// See: https://discord.com/developers/docs/topics/threads#gaining-access-to-a-channel
+func (s *State) ThreadListSync(t *ThreadListSync) error {
+	if s == nil {
+		return ErrNilState
+	}
+
+	g, err := s.Guild(t.GuildID)
+	if err != nil {
+		return err
+	}
+
+	// if no channel IDs were given, all channel IDs were already synchronized
+	if len(t.ChannelIDs) != 0 {
+
+		s.Lock()
+		defer s.Unlock()
+
+		// All channel IDs provided are the parent channel IDs that we need to synchronize.
+		// The Discord API tells us that we should use this info to clear out all active threads whose
+		// parent channel IDs are specified.
+		for _, parentChannelID := range t.ChannelIDs {
+			for _, stateChannel := range s.channelMap {
+				if stateChannel.ParentID == parentChannelID && (stateChannel.Type == ChannelTypeGuildPublicThread || stateChannel.Type == ChannelTypeGuildPrivateThread || stateChannel.Type == ChannelTypeGuildNewsThread) && !stateChannel.ThreadMetadata.Archived {
+					for i, c := range g.Channels {
+						if c.ID == stateChannel.ID {
+							g.Channels = append(g.Channels[:i], g.Channels[i+1:]...)
+							break
+						}
+					}
+					delete(s.channelMap, stateChannel.ID)
+				}
+			}
+		}
+
+		// Ingest new threads given to us by the sync event. This will contain all NEW active threads
+		// that our client can see.
+		for _, channel := range t.Threads {
+			s.channelMap[channel.ID] = &channel
+			g.Channels = append(g.Channels, &channel)
+		}
+
+		// Set the member field of the threads based on the list of thread member data given to us by the sync event.
+		for _, member := range t.Members {
+			if channel, ok := s.channelMap[member.ID]; ok {
+				channel.Member = &member
+			}
+		}
+
+		return nil
+	}
+
+	return ErrStateNotFound
+}
```

Comment:
This is copied from https://github.com/bwmarrin/discordgo/pull/922/commits/d890873d60f2958fa010609793c483ed2d0ab59e#diff-320959764bfa2277e0bdd4eaa741f02891d123c42150d4ccb76de0bdbf6e591aR606-R650, and includes many of the same idiosyncrasies—the nested loops which take a very unusual pattern (I cannot imagine two people working independently coming up with this same strategy), the same issue of misinterpreting the case of `len(t.ChannelIDs) == 0`, and the same returning of `ErrStateNotFound` in a situation where it really doesn't make sense to return that error.